File: Scrabbler.Tests/BoardWordExtractorTests.cs
using Scrabbler.Domain.BoardModel;

namespace Scrabbler.Tests;

public sealed class BoardWordExtractorTests
{
    [Fact]
    public void ExtractsHorizontalAndVerticalWordsWithCoordinates()
    {
        var board = EmptyBoard()
            .SetCell(7, 7, 'K')
            .SetCell(7, 8, 'O')
            .SetCell(7, 9, 'T')
            .SetCell(6, 8, 'N')
            .SetCell(8, 8, 'S');

        var words = BoardWordExtractor.ExtractWords(board);

        var horizontal = Assert.Single(words, word => word.Direction == Direction.Horizontal);
        Assert.Equal("KOT", horizontal.Text);
        Assert.Equal([(7, 7), (7, 8), (7, 9)], horizontal.Cells.Select(cell => (cell.Row, cell.Column)).ToArray());

        var vertical = Assert.Single(words, word => word.Direction == Direction.Vertical);
        Assert.Equal("NOS", vertical.Text);
        Assert.Equal([(6, 8), (7, 8), (8, 8)], vertical.Cells.Select(cell => (cell.Row, cell.Column)).ToArray());
    }

    [Fact]
    public void IgnoresSingleLetterRuns()
    {
        var board = EmptyBoard()
            .SetCell(0, 0, 'A')
            .SetCell(2, 2, 'K')
            .SetCell(2, 3, 'O');

        var words = BoardWordExtractor.ExtractWords(board);

        var word = Assert.Single(words);
        Assert.Equal("KO", word.Text);
    }

    private static Board EmptyBoard()
    {
        return new Board(new BonusType[Board.Size, Board.Size]);
    }
}

File: Scrabbler.iOS/ScrabblerKit/Tests/ScrabblerKitTests/BoardWordExtractorTests.swift
import Testing
@testable import ScrabblerKit

@Suite("Board word extractor")
struct BoardWordExtractorTests {
    @Test func extractsHorizontalAndVerticalWordsWithCoordinates() {
        let board = emptyBoard()
            .setCell(row: 0, column: 0, letter: "A")
            .setCell(row: 0, column: 1, letter: "L")
            .setCell(row: 0, column: 2, letter: "A")
            .setCell(row: 1, column: 0, letter: "S")
            .setCell(row: 2, column: 0, letter: "A")

        let words = BoardWordExtractor.extractWords(from: board)

        #expect(words.contains { $0.text == "ALA" && $0.direction == .horizontal && $0.coordinates.map(\.column) == [0, 1, 2] })
        #expect(words.contains { $0.text == "ASA" && $0.direction == .vertical && $0.coordinates.map(\.row) == [0, 1, 2] })
    }

    @Test func ignoresSingleLetterRuns() {
        let board = emptyBoard().setCell(row: 0, column: 0, letter: "A")

        #expect(BoardWordExtractor.extractWords(from: board).isEmpty)
    }

    private func emptyBoard() -> Board {
        Board(bonuses: Array(repeating: Array(repeating: BonusType.none, count: Board.size), count: Board.size))
    }
}
